File: android/src/main/kotlin/io/adrop/adrop_ads/consent/FlutterAdropConsentManager.kt
package io.adrop.adrop_ads.consent

import android.app.Activity
import android.content.Context
import android.provider.Settings
import io.adrop.ads.Adrop
import io.adrop.ads.consent.AdropConsentDebugGeography
import io.adrop.ads.consent.AdropConsentListener
import io.adrop.ads.consent.AdropConsentResult
import io.adrop.ads.model.AdropErrorCode
import io.flutter.plugin.common.MethodChannel.Result
import java.security.MessageDigest

class FlutterAdropConsentManager {

    fun requestConsentInfoUpdate(activity: Activity?, result: Result) {
        val consentManager = Adrop.consentManager
        if (consentManager == null) {
            result.error(
                AdropErrorCode.ERROR_CODE_INTERNAL.name,
                "ConsentManager is not available",
                "AdropAdsBackfill module is not installed"
            )
            return
        }
        if (activity == null) {
            result.error(
                AdropErrorCode.ERROR_CODE_INITIALIZE.name,
                "Activity is not available",
                null
            )
            return
        }
        consentManager.requestConsentInfoUpdate(activity, object : AdropConsentListener {
            override fun onConsentInfoUpdated(consentResult: AdropConsentResult) {
                val resultMap = hashMapOf<String, Any?>(
                    "status" to consentResult.status.value,
                    "canRequestAds" to consentResult.canRequestAds,
                    "canShowPersonalizedAds" to consentResult.canShowPersonalizedAds,
                    "error" to consentResult.error?.message
                )
                result.success(resultMap)
            }
        })
    }

    fun getConsentStatus(context: Context?, result: Result) {
        val consentManager = Adrop.consentManager
        if (consentManager == null || context == null) {
            result.success(0) // UNKNOWN
            return
        }
        val status = consentManager.getConsentStatus(context)
        result.success(status.value)
    }

    fun canRequestAds(context: Context?, result: Result) {
        val consentManager = Adrop.consentManager
        if (consentManager == null || context == null) {
            result.success(false)
            return
        }
        val canRequest = consentManager.canRequestAds(context)
        result.success(canRequest)
    }

    fun reset(context: Context?, result: Result) {
        val consentManager = Adrop.consentManager
        if (consentManager == null || context == null) {
            result.success(null)
            return
        }
        consentManager.reset(context)
        result.success(null)
    }

    fun setDebugSettings(context: Context?, geographyValue: Int, result: Result) {
        val consentManager = Adrop.consentManager
        if (consentManager == null || context == null) {
            result.success(null)
            return
        }

        val androidId = Settings.Secure.getString(
            context.contentResolver,
            Settings.Secure.ANDROID_ID
        )
        val hash = MessageDigest.getInstance("MD5")
            .digest(androidId.toByteArray())
            .joinToString("") { "%02X".format(it) }

        val geography = when (geographyValue) {
            0 -> AdropConsentDebugGeography.DISABLED
            1 -> AdropConsentDebugGeography.EEA
            3 -> AdropConsentDebugGeography.REGULATED_US_STATE
            4 -> AdropConsentDebugGeography.OTHER
            else -> AdropConsentDebugGeography.DISABLED
        }

        consentManager.setDebugSettings(listOf(hash), geography)
        result.success(null)
    }
}

File: ios/Classes/consent/FlutterAdropConsentManager.swift
import Flutter
import UIKit
import AdropAds

class FlutterAdropConsentManager {

    func requestConsentInfoUpdate(result: @escaping FlutterResult) {
        guard let consentManager = Adrop.consentManager else {
            result(FlutterError(
                code: "ERROR_CODE_INTERNAL",
                message: "ConsentManager is not available",
                details: "AdropAdsBackfill module is not installed"
            ))
            return
        }

        guard let viewController = getTopViewController() else {
            result(FlutterError(
                code: "ERROR_CODE_INTERNAL",
                message: "Top view controller is not available",
                details: "Unable to retrieve a valid UIViewController to present consent UI"
            ))
            return
        }

        consentManager.requestConsentInfoUpdate(from: viewController) { consentResult in
            let resultMap: [String: Any?] = [
                "status": consentResult.status.rawValue,
                "canRequestAds": consentResult.canRequestAds,
                "canShowPersonalizedAds": consentResult.canShowPersonalizedAds,
                "error": consentResult.error?.localizedDescription
            ]
            result(resultMap)
        }
    }

    func getConsentStatus(result: @escaping FlutterResult) {
        guard let consentManager = Adrop.consentManager else {
            result(0) // UNKNOWN
            return
        }
        let status = consentManager.getConsentStatus()
        result(status.rawValue)
    }

    func canRequestAds(result: @escaping FlutterResult) {
        guard let consentManager = Adrop.consentManager else {
            result(false)
            return
        }
        let canRequest = consentManager.canRequestAds()
        result(canRequest)
    }

    func reset(result: @escaping FlutterResult) {
        guard let consentManager = Adrop.consentManager else {
            result(nil)
            return
        }
        consentManager.reset()
        result(nil)
    }

    func setDebugSettings(geographyValue: Int, result: @escaping FlutterResult) {
        guard let consentManager = Adrop.consentManager else {
            result(nil)
            return
        }

        let geography: AdropConsentDebugGeography
        switch geographyValue {
        case 0:
            geography = .disabled
        case 1:
            geography = .EEA
        case 3:
            geography = .regulatedUSState
        case 4:
            geography = .other
        default:
            geography = .disabled
        }

        let deviceIdentifier = UIDevice.current.identifierForVendor?.uuidString ?? ""
        consentManager.setDebugSettings(testDeviceIdentifiers: [deviceIdentifier], geography: geography)
        result(nil)
    }

    private func getTopViewController() -> UIViewController? {
        var keyWindow: UIWindow?
        if #available(iOS 13.0, *) {
            keyWindow = UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .flatMap { $0.windows }
                .first { $0.isKeyWindow }
        } else {
            keyWindow = UIApplication.shared.keyWindow
        }
        return keyWindow?.rootViewController
    }
}
